book/src/developers/designs/application_passwords.md

```diff
@@ -0,0 +1,188 @@
+but then still need other groups reflected into the subtree. In this case, if we
+limited the view to application access only, we wouldn't be displaying the
+non-access groups that the application may still rely on. Ultimately, in this
+case the application needs to make it's own authorisation decisions to an
+extend. Kanidm can limit which users are members of the access allowed group as
+only they can bind still as an extra layer of defence)
+
+The application must bind with its token if it wishes to read extended user
+information. With this, only basic info limited to anonymous rights are granted.
+
+(NOTE: We can't assume these DNs are private - I did consider making these
+app=<secret key>,dc=example,dc=com, but client applications may disclose this
+basedn in ui elements).
+
+When a user authenticates the binddn of the account is set to
+spn=user@label,app=name,dc=example,dc=com. This difference in base DN triggers
+Kanidm to re-route the authentication to the application specific password,
+rather than the UNIX one.
+
+## Kanidm
+
+### Application Entries
+
+A new class for applications will be added that has an associated reference to a
+set of groups that are allowed access.
+
+(NOTE: It could be good to make this a group such that members of the group
+inherit "memberof=application X". This would make presentation of application
+password links in the UI easier since we can then do a search on (memberof &&
+class = application). Alternately we don't need this since we can in-memory
+cache applications and just compare their member sets to the user sessions
+memberof)
+
+Applications need to have service-account like properties where tokens can be
+generated for them. These are optional since an anonymous bind to kanidm and
+searching under the basedn will continue to work.
+
+Application should have a URL reference to help admins identify where the
+application may be located or accessed.
+
+(NOTE: Future, it could be good to allow customisable instructions for users on
+where to go to use their app password?)
+
+(NOTE: To make it easier to manage tokens, it could be good to have a
+"token-capable" class type to prevent duplication this with service-accounts,
+and allows adding tokens to future types).
+
+### Accounts
+
+The user may wish to have multiple passwords per application. Each password may
+have a label to identify it. For example:
+
+```
+  MAIL
+    iphone: abcd...
+    laptop: bcde...
+  HTTP
+    workstation: cdef...
+```
+
+Person accounts will need a new `Attribute::ApplicationPassword` that stores a
+`ValueSetApplicationPassword`. Each value in the set is a new type to manage these
+secrets and their labeling and the references to the applications.
+
+```
+struct ApplicationPassword {
+  uuid: Uuid,
+  application: Uuid,
+  label: String,
+  password: Password,
+}
+
+struct ValueSetApplicationPassword {
+  map: Map<Uuid, ApplicationPassword>
+}
+```
+
```

Comment:
I think this may be inverted.

You need ValueSetApplicationPassword to be a map of:

```

type ApplicationPasswords = BTreeMap<Uuid, (String, Password)>;

type ApplicationUuid = Uuid;

struct ValueSetApplicationPassword {
  map: Map<ApplicationUuid, ApplicationPasswords>
}
```

This way when the authentication request comes in for app=X, we can get the applications UUID in the first step. This allows us to look up the account and then get application passwords by the UUID of the app. 

In the design you propose, we would need to iterate over all possible application passwords to find the one that matches rather than being limited to getting just the ones related to the application ID.

Reply:
Ok, updated. It will impact queries by Uuid `PartialValue::ApplicationPassword(Uuid)` but those should be less frequent than authentication requests.

Reply:
I doubt those are going to be the common thing though. We should be optimising for the common case here.